File: src/robot_hardware_bridge/robot_hardware_bridge/bridge_node.py
"""
Hardware Bridge Node

AUTHORITY: RELAY ONLY - NO CONTROL AUTHORITY

This node acts as a protocol gateway between ROS 2 and the RT Controller.
It performs SCHEMA validation only; all semantic validation is done by Controller.

Responsibilities:
- Subscribe to trajectory commands from MoveIt
- Validate message schema (types, field presence, array lengths)
- Serialize and transmit to Controller (or Gazebo backend in SIM)
- Receive joint states from Controller and publish to ROS
- Receive TrajectoryAck and ExecutionState from Controller

Explicit Non-Responsibilities:
- Joint limit checking (Controller does this)
- Velocity bound checking (Controller does this)
- Safety decisions (Controller does this)
- Motor control (Controller does this)
"""

import os
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, HistoryPolicy

from sensor_msgs.msg import JointState
from trajectory_msgs.msg import JointTrajectory, JointTrajectoryPoint
from robot_msgs.msg import TrajectoryAck, ExecutionState

# Import backends
from .sim_backend import SimulationBackend
from .real_backend import RealControllerBackend


class HardwareBridgeNode(Node):
    """
    Hardware Bridge Node - RELAY ONLY
    
    This node relays commands between ROS and the Controller.
    It has NO authority over motor control or safety.
    """
    
    # Expected joint names (must match URDF)
    JOINT_NAMES = ['joint_1', 'joint_2', 'joint_3', 'joint_4', 'joint_5', 'joint_6']
    
    def __init__(self):
        super().__init__('hardware_bridge')
        
        # Declare parameters
        self.declare_parameter('robot_mode', 'SIM')
        self.declare_parameter('controller_type', 'FAKE')
        self.declare_parameter('controller_ip', '192.168.1.100')
        self.declare_parameter('controller_port', 5000)
        self.declare_parameter('controller_proto', 'TCP')
        
        # Get parameters
        self.robot_mode = self.get_parameter('robot_mode').get_parameter_value().string_value
        self.controller_type = self.get_parameter('controller_type').get_parameter_value().string_value
        self.controller_ip = self.get_parameter('controller_ip').get_parameter_value().string_value
        self.controller_port = self.get_parameter('controller_port').get_parameter_value().integer_value
        self.controller_proto = self.get_parameter('controller_proto').get_parameter_value().string_value
        
        # Override from environment if present
        self.robot_mode = os.environ.get('ROBOT_MODE', self.robot_mode)
        self.controller_type = os.environ.get('CONTROLLER_TYPE', self.controller_type)
        self.controller_ip = os.environ.get('CONTROLLER_IP', self.controller_ip)
        self.controller_port = int(os.environ.get('CONTROLLER_PORT', str(self.controller_port)))
        self.controller_proto = os.environ.get('CONTROLLER_PROTO', self.controller_proto)
        
        self.get_logger().info(f'Hardware Bridge starting in {self.robot_mode} mode')
        self.get_logger().info(f'Controller type: {self.controller_type}')
        
        # Initialize appropriate backend
        if self.robot_mode == 'SIM' or self.controller_type == 'FAKE' or self.controller_type == 'GAZEBO':
            self.get_logger().info(f'Using Simulation Backend ({self.controller_type} Mode)')
            self.backend = SimulationBackend(self, self.controller_type)
        else:
            self.get_logger().info(f'Using Real Controller Backend at {self.controller_ip}:{self.controller_port}')
            self.backend = RealControllerBackend(
                self,
                ip=self.controller_ip,
                port=self.controller_port,
                protocol=self.controller_proto
            )
        
        # QoS for reliable trajectory delivery
        trajectory_qos = QoSProfile(
            reliability=ReliabilityPolicy.RELIABLE,
            history=HistoryPolicy.KEEP_LAST,
            depth=10
        )
        
        # QoS for best-effort joint state streaming
        joint_state_qos = QoSProfile(
            reliability=ReliabilityPolicy.BEST_EFFORT,
            history=HistoryPolicy.KEEP_LAST,
            depth=1
        )
        
        # === SUBSCRIBERS (Commands from ROS) ===
        self.trajectory_sub = self.create_subscription(
            JointTrajectory,
            '/planned_trajectory',
            self.trajectory_callback,
            trajectory_qos
        )
        
        # === PUBLISHERS (State to ROS) ===
        self.joint_state_pub = self.create_publisher(
            JointState,
            '/joint_states',
            joint_state_qos
        )
        
        self.trajectory_ack_pub = self.create_publisher(
            TrajectoryAck,
            '/trajectory_ack',
            trajectory_qos
        )
        
        self.execution_state_pub = self.create_publisher(
            ExecutionState,
            '/execution_state',
            trajectory_qos
        )
        
        # Start backend communication
        self.backend.start()
        
        # Timer for state publishing
        self.create_timer(0.01, self.publish_state_callback)  # 100 Hz
        
        self.get_logger().info('Hardware Bridge initialized')
    
    def trajectory_callback(self, msg: JointTrajectory):
        """
        Handle incoming trajectory from MoveIt.
        
        Performs SCHEMA validation only:
        - Check message type
        - Check joint names match expected
        - Check points array is not empty
        - Check array lengths are consistent
        
        Does NOT perform semantic validation (Controller does that):
        - Joint limits
        - Velocity bounds
        - Physics constraints
        """
        self.get_logger().info(f'Received trajectory with {len(msg.points)} points')
        
        # === SCHEMA VALIDATION ONLY ===
        
        # Validate joint names match expected
        if len(msg.joint_names) != len(self.JOINT_NAMES):
            self.get_logger().error(f'Schema error: Expected {len(self.JOINT_NAMES)} joints, got {len(msg.joint_names)}')
            self._publish_rejection(msg, 'SCHEMA_ERROR_JOINT_COUNT')
            return
        
        for i, name in enumerate(msg.joint_names):
            if name != self.JOINT_NAMES[i]:
                self.get_logger().error(f'Schema error: Expected joint {self.JOINT_NAMES[i]}, got {name}')
                self._publish_rejection(msg, f'SCHEMA_ERROR_JOINT_NAME_{i}')
                return
        
        # Validate points array is not empty
        if len(msg.points) == 0:
            self.get_logger().error('Schema error: Empty points array')
            self._publish_rejection(msg, 'SCHEMA_ERROR_EMPTY_POINTS')
            return
        
        # Validate each point has correct array lengths
        for i, point in enumerate(msg.points):
            if len(point.positions) != len(self.JOINT_NAMES):
                self.get_logger().error(f'Schema error: Point {i} has {len(point.positions)} positions, expected {len(self.JOINT_NAMES)}')
                self._publish_rejection(msg, f'SCHEMA_ERROR_POSITIONS_LENGTH_{i}')
                return
        
        # Schema validation passed - relay to backend
        # NOTE: Semantic validation (limits, velocities) is done by Controller
        self.get_logger().debug('Schema validation passed, relaying to backend')
        self.backend.send_trajectory(msg)
    
    def _publish_rejection(self, msg: JointTrajectory, reason: str):
        """Publish trajectory rejection due to schema error."""
        ack = TrajectoryAck()
        ack.trajectory_id = msg.header.stamp
        ack.status = TrajectoryAck.STATUS_REJECTED
        ack.reject_reason = reason
        self.trajectory_ack_pub.publish(ack)
    
    def publish_state_callback(self):
        """Publish current state from backend to ROS."""
        # Get joint state from backend
        joint_state = self.backend.get_joint_state()
        if joint_state is not None:
            self.joint_state_pub.publish(joint_state)
        
        # Get execution state from backend
        exec_state = self.backend.get_execution_state()
        if exec_state is not None:
            self.execution_state_pub.publish(exec_state)
        
        # Get any pending trajectory acks
        ack = self.backend.get_trajectory_ack()
        if ack is not None:
            self.trajectory_ack_pub.publish(ack)
    
    def destroy_node(self):
        """Clean shutdown."""
        self.get_logger().info('Shutting down Hardware Bridge')
        self.backend.stop()
        super().destroy_node()


def main(args=None):
    rclpy.init(args=args)
    node = HardwareBridgeNode()
    
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/robot_hardware_bridge/src/bridge_node.cpp
/**
 * Hardware Bridge Node (C++ Version)
 * 
 * AUTHORITY: RELAY ONLY - NO CONTROL AUTHORITY
 * 
 * C++ implementation for deterministic timing and real-time performance.
 */

#include "robot_hardware_bridge/bridge_node.hpp"
#include <cstdlib>
#include <chrono>

using namespace std::chrono_literals;

namespace robot_hardware_bridge
{

// Static member definition
const std::vector<std::string> HardwareBridgeNode::JOINT_NAMES = {
  "joint_1", "joint_2", "joint_3", "joint_4", "joint_5", "joint_6"
};

// ============================================================================
// SimulationBackend Implementation
// ============================================================================

SimulationBackend::SimulationBackend(rclcpp::Node* node, const std::string& mode)
  : node_(node), mode_(mode)
{
  current_positions_.resize(NUM_JOINTS, 0.0);
  target_positions_.resize(NUM_JOINTS, 0.0);
}

void SimulationBackend::start()
{
  RCLCPP_INFO(node_->get_logger(), "Simulation backend started in %s mode", mode_.c_str());
}

void SimulationBackend::stop()
{
  RCLCPP_INFO(node_->get_logger(), "Simulation backend stopped");
}

void SimulationBackend::send_trajectory(const trajectory_msgs::msg::JointTrajectory& msg)
{
  if (msg.points.empty()) {
    return;
  }
  
  std::lock_guard<std::mutex> lock(state_mutex_);
  
  // Set target to final point
  const auto& final_point = msg.points.back();
  for (size_t i = 0; i < NUM_JOINTS && i < final_point.positions.size(); ++i) {
    target_positions_[i] = final_point.positions[i];
  }
  
  executing_ = true;
  
  // Create pending ack
  pending_ack_ = std::make_shared<robot_msgs::msg::TrajectoryAck>();
  pending_ack_->trajectory_id = msg.header.stamp;
  pending_ack_->status = robot_msgs::msg::TrajectoryAck::STATUS_ACCEPTED;
  
  RCLCPP_INFO(node_->get_logger(), "Trajectory accepted, moving to target");
}

sensor_msgs::msg::JointState::SharedPtr SimulationBackend::get_joint_state()
{
  std::lock_guard<std::mutex> lock(state_mutex_);
  
  // Interpolate towards target
  if (executing_) {
    bool reached = true;
    for (size_t i = 0; i < NUM_JOINTS; ++i) {
      double diff = target_positions_[i] - current_positions_[i];
      if (std::abs(diff) > 0.001) {
        current_positions_[i] += diff * INTERPOLATION_RATE;
        reached = false;
      } else {
        current_positions_[i] = target_positions_[i];
      }
    }
    if (reached) {
      executing_ = false;
      RCLCPP_INFO(node_->get_logger(), "Target position reached");
    }
  }
  
  auto msg = std::make_shared<sensor_msgs::msg::JointState>();
  msg->header.stamp = node_->now();
  msg->name = {"joint_1", "joint_2", "joint_3", "joint_4", "joint_5", "joint_6"};
  msg->position = current_positions_;
  msg->velocity.resize(NUM_JOINTS, 0.0);
  msg->effort.resize(NUM_JOINTS, 0.0);
  
  return msg;
}

robot_msgs::msg::ExecutionState::SharedPtr SimulationBackend::get_execution_state()
{
  auto msg = std::make_shared<robot_msgs::msg::ExecutionState>();
  msg->header.stamp = node_->now();
  msg->is_executing = executing_.load();
  msg->progress = executing_ ? 0.5 : 1.0;
  return msg;
}

robot_msgs::msg::TrajectoryAck::SharedPtr SimulationBackend::get_trajectory_ack()
{
  std::lock_guard<std::mutex> lock(state_mutex_);
  auto ack = pending_ack_;
  pending_ack_.reset();
  return ack;
}

// ============================================================================
// HardwareBridgeNode Implementation
// ============================================================================

HardwareBridgeNode::HardwareBridgeNode()
  : Node("hardware_bridge_cpp")
{
  // Declare parameters
  this->declare_parameter("robot_mode", "SIM");
  this->declare_parameter("controller_type", "FAKE");
  this->declare_parameter("controller_ip", "192.168.1.100");
  this->declare_parameter("controller_port", 5000);
  this->declare_parameter("controller_proto", "TCP");
  
  // Get parameters
  robot_mode_ = this->get_parameter("robot_mode").as_string();
  controller_type_ = this->get_parameter("controller_type").as_string();
  controller_ip_ = this->get_parameter("controller_ip").as_string();
  controller_port_ = this->get_parameter("controller_port").as_int();
  controller_proto_ = this->get_parameter("controller_proto").as_string();
  
  // Override from environment
  if (const char* env = std::getenv("ROBOT_MODE")) {
    robot_mode_ = env;
  }
  if (const char* env = std::getenv("CONTROLLER_TYPE")) {
    controller_type_ = env;
  }
  
  RCLCPP_INFO(this->get_logger(), "Hardware Bridge (C++) starting in %s mode", robot_mode_.c_str());
  RCLCPP_INFO(this->get_logger(), "Controller type: %s", controller_type_.c_str());
  
  // Initialize backend (simulation only for now)
  backend_ = std::make_unique<SimulationBackend>(this, controller_type_);
  
  // QoS profiles
  auto trajectory_qos = rclcpp::QoS(10).reliable();
  auto joint_state_qos = rclcpp::QoS(1).best_effort();
  
  // Publishers
  joint_state_pub_ = this->create_publisher<sensor_msgs::msg::JointState>(
    "/joint_states", joint_state_qos);
  
  trajectory_ack_pub_ = this->create_publisher<robot_msgs::msg::TrajectoryAck>(
    "/trajectory_ack", trajectory_qos);
  
  execution_state_pub_ = this->create_publisher<robot_msgs::msg::ExecutionState>(
    "/execution_state", trajectory_qos);
  
  // Subscribers
  trajectory_sub_ = this->create_subscription<trajectory_msgs::msg::JointTrajectory>(
    "/planned_trajectory",
    trajectory_qos,
    std::bind(&HardwareBridgeNode::trajectory_callback, this, std::placeholders::_1));
  
  // Start backend
  backend_->start();
  
  // State publishing timer (100 Hz for real-time performance)
  state_timer_ = this->create_wall_timer(
    10ms,  // 100 Hz
    std::bind(&HardwareBridgeNode::publish_state_callback, this));
  
  RCLCPP_INFO(this->get_logger(), "Hardware Bridge (C++) initialized");
}

HardwareBridgeNode::~HardwareBridgeNode()
{
  RCLCPP_INFO(this->get_logger(), "Shutting down Hardware Bridge (C++)");
  if (backend_) {
    backend_->stop();
  }
}

void HardwareBridgeNode::trajectory_callback(
  const trajectory_msgs::msg::JointTrajectory::SharedPtr msg)
{
  RCLCPP_INFO(this->get_logger(), "Received trajectory with %zu points", msg->points.size());
  
  // Schema validation
  if (!validate_schema(msg)) {
    return;
  }
  
  // Relay to backend
  RCLCPP_DEBUG(this->get_logger(), "Schema validation passed, relaying to backend");
  backend_->send_trajectory(*msg);
}

bool HardwareBridgeNode::validate_schema(
  const trajectory_msgs::msg::JointTrajectory::SharedPtr msg)
{
  // Validate joint count
  if (msg->joint_names.size() != JOINT_NAMES.size()) {
    RCLCPP_ERROR(this->get_logger(), 
      "Schema error: Expected %zu joints, got %zu",
      JOINT_NAMES.size(), msg->joint_names.size());
    publish_rejection(msg, "SCHEMA_ERROR_JOINT_COUNT");
    return false;
  }
  
  // Validate joint names
  for (size_t i = 0; i < JOINT_NAMES.size(); ++i) {
    if (msg->joint_names[i] != JOINT_NAMES[i]) {
      RCLCPP_ERROR(this->get_logger(),
        "Schema error: Expected joint %s, got %s",
        JOINT_NAMES[i].c_str(), msg->joint_names[i].c_str());
      publish_rejection(msg, "SCHEMA_ERROR_JOINT_NAME");
      return false;
    }
  }
  
  // Validate points not empty
  if (msg->points.empty()) {
    RCLCPP_ERROR(this->get_logger(), "Schema error: Empty points array");
    publish_rejection(msg, "SCHEMA_ERROR_EMPTY_POINTS");
    return false;
  }
  
  // Validate position array lengths
  for (size_t i = 0; i < msg->points.size(); ++i) {
    if (msg->points[i].positions.size() != JOINT_NAMES.size()) {
      RCLCPP_ERROR(this->get_logger(),
        "Schema error: Point %zu has %zu positions, expected %zu",
        i, msg->points[i].positions.size(), JOINT_NAMES.size());
      publish_rejection(msg, "SCHEMA_ERROR_POSITIONS_LENGTH");
      return false;
    }
  }
  
  return true;
}

void HardwareBridgeNode::publish_rejection(
  const trajectory_msgs::msg::JointTrajectory::SharedPtr msg,
  const std::string& reason)
{
  auto ack = robot_msgs::msg::TrajectoryAck();
  ack.trajectory_id = msg->header.stamp;
  ack.status = robot_msgs::msg::TrajectoryAck::STATUS_REJECTED;
  ack.reject_reason = reason;
  trajectory_ack_pub_->publish(ack);
}

void HardwareBridgeNode::publish_state_callback()
{
  // Publish joint state
  auto joint_state = backend_->get_joint_state();
  if (joint_state) {
    joint_state_pub_->publish(*joint_state);
  }
  
  // Publish execution state
  auto exec_state = backend_->get_execution_state();
  if (exec_state) {
    execution_state_pub_->publish(*exec_state);
  }
  
  // Publish any pending trajectory acks
  auto ack = backend_->get_trajectory_ack();
  if (ack) {
    trajectory_ack_pub_->publish(*ack);
  }
}

}  // namespace robot_hardware_bridge

// Main entry point
int main(int argc, char* argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<robot_hardware_bridge::HardwareBridgeNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
